### ISSUE TITLE
Accessing JTI in storage for a JWT

### ISSUE THREAD

**Issue by cristiangraz** [ory/fosite#122]
The storage receives the token's signature, but doesn't have the ability to access the JTI for the JWT. Would you be opposed to a PR to make the JTI accessible to storage? I was thinking the `JWTSessionContainer` could include two additional interface methods:

``` go
// SetJTI sets the JTI for the given token type.
SetJTI(fosite.TokenType, string)

// GetJTI retrieves the JTI for the given token type.
GetJTI(fosite.TokenType) string
```

The `generate` method in `R256JWTStrategy` could handle setting the JTI and storing in the session before calling `ToMapClaims`. Storage could then access with `session.GetJTI(fosite.AccessToken)`

**Comment by arekkas** [comment]
I think that you can achieve this if you use your own session object pretty easily. I don't think it is requried to add this to the fosite API, as JTI is not part of the mandatory spec.

**Comment by cristiangraz** [comment]
I don't think a custom session object alone is enough.

If you set the JTI initially in the `JWTClaims` of the session, you can access in storage -- but only if you are _only_ issuing an access token. If you issue a refresh token they will both have the same JTI as the JTI gets reused if it already exists (see link below).

If you depend on the JTI generated in `RS256JWTStrategy` so that the JTIs are unique, it's generated and assigned to a local map to be signed and is not accessible to the storage.

See here: https://github.com/ory-am/fosite/blob/cb328caca6287c7995ee5285c6446bffd4ef496b/token/jwt/claims_jwt.go#L24-L29

Am I missing something in terms of how to access the unique JTI in storage?

There's some scenarios where having the JWI is helpful, like avoiding db lookups after cryptographic verification but checking for an in-memory blacklist of the JTI, or using the JTI as an alternate way to revoke a token. Because the JWT implementation has it's own session object, and JTI is part of that, I thought it would make sense to include in fosite.

Either way, this can be implemented by creating a custom `CoreStrategy` struct that embeds `oauth2.RS256JWTStrategy` and implements a few wrapper functions.

**Comment by arekkas** [comment]
> If you set the JTI initially in the JWTClaims of the session, you can access in storage -- but only if you are only issuing an access token. If you issue a refresh token they will both have the same JTI as the JTI gets reused if it already exists (see link below).

Ah I see, that makes sense, of course. And it's actually a bug, because each token should have it's unique JTI. I would like to fix this without breaking backwards compatibility, do you have any ideas?

**Comment by cristiangraz** [comment]
I think the easiest solution right now to avoid JTIs being reused would be to generate a new UUID each time `ToMap` is called and ignoring any JTI set on the JWTClaims. 

Re BC the only issue I can think of would be someone who is currently setting the JTI in their session and relying on that in storage. Was that the intent behind using the existing JTI in the `ToMap` function linked above? Or if someone wanted to implement their own custom ID generator?

From a BC perspective, I'm not sure many people are doing this just given the bug mentioned above with the same JTI for access and refresh tokens.

**Comment by cristiangraz** [comment]
It seems like the idea of putting a single `JWTClaims` in the shared session may be a central problem given that they are reused for each token but have differences (like the `ExpiresAt` map you recently added for `0.5.0`, and now the JTI). Other limitations might be the ability to implement claims per token type (i.e. if you wanted to differentiate refresh token from access token so you could rely strictly on cryptographic verification but prevent refresh tokens from being used as access tokens). An alternate approach (although not BC) might be switching from the current JWT session setup:

```go
return &Session{
    JWTSession: &oauth2.JWTSession{
        JWTClaims: &jwt.JWTClaims{
            Issuer:    issuer,
            IssuedAt:  now,
            NotBefore: now,
            Extra:     make(map[string]interface{}),
        },
        ExpiresAt: map[fosite.TokenType]time.Time{
            fosite.AuthorizeCode: now.Add(config.AuthorizeCodeLifespan),
            fosite.AccessToken:   now.Add(config.AccessTokenLifespan),
            fosite.RefreshToken:  now.Add(config.RefreshTokenLifespan),
        },
    },
}
```
to something like this:

```go
now := time.Now()
ac, at, rt := uuid.new(), uuid.New(), uuid.New()
return &Session{
    JWTSession: &oauth2.JWTSession{
        Issuer:    issuer,
        IssuedAt:  now,
        NotBefore: now,
        AuthorizeCode: &jwt.PartialClaims{
            JTI: ac,
            ExpiresIn: config.AuthorizeCodeLifespan,
            Extra:     make(map[string]interface{}),
        },
        AccessToken: &jwt.PartialClaims{
            JTI: at,
            ExpiresIn: config.AccessTokenLifespan,
            Extra:     make(map[string]interface{}),
        },
        RefreshToken: &jwt.PartialClaims{
            JTI: rt,
            ExpiresIn: config.RefreshTokenLifespan,
            Extra:     map[string]interface{}{
                "ati": at, // access token id to indicate this is a refresh token
            },
        },
    },
}
```
Then you could rely on the existing JTI if set, or if not, generate a UUID in the `ToMap` function.

You can achieve the same thing by creating a minimal `CoreStrategy` wrapper, so not sure this would all be necessary. Just putting some thoughts out there.

**Comment by arekkas** [comment]
That could be one possibility, I will think about this more tomorrow and let you know if I come up with any alternatives :)

**Comment by arekkas** [comment]
Sorry, I completely forgot about this. Are you still down for this @cristiangraz ?

**Comment by arekkas** [comment]
I like the first approach, because it is more dense:

```go
return &Session{
    JWTSession: &oauth2.JWTSession{
        JWTClaims: &jwt.JWTClaims{
            Issuer:    issuer,
            IssuedAt:  now,
            NotBefore: now,
            Extra:     make(map[string]interface{}),
        },
        ExpiresAt: map[fosite.TokenType]time.Time{
            fosite.AuthorizeCode: now.Add(config.AuthorizeCodeLifespan),
            fosite.AccessToken:   now.Add(config.AccessTokenLifespan),
            fosite.RefreshToken:  now.Add(config.RefreshTokenLifespan),
        },
    },
}
```

The JWT Strategy would then be generating the JTIs uniquely

**Comment by arekkas** [comment]
Would you mind taking a shot @cristiangraz ?

**Comment by cristiangraz** [comment]
@arekkas Just to confirm, it's this:

> I think the easiest solution right now to avoid JTIs being reused would be to generate a new UUID each time ToMap is called and ignoring any JTI set on the JWTClaims.

**Comment by arekkas** [comment]
I reconsidered, I like your initial approach with

```
// SetJTI sets the JTI for the given token type.
SetJTI(fosite.TokenType, string)

// GetJTI retrieves the JTI for the given token type.
GetJTI(fosite.TokenType) string
```

the best one. Even if it breaks bc it's fine, as we're not in 1.0.0 yet. :)

**Comment by cristiangraz** [comment]
Ah okay. I have a working implementation I can port to Fosite, will try to get a PR in later this week.

**Comment by arekkas** [comment]
awesome!

**Comment by petejkim** [comment]
Is this still being worked on?

**Comment by arekkas** [comment]
This will be resolved with #180 because refresh and authorize codes will no longer be JWT-based, but hmac.
